gr-qtgui/lib/matrix_display.cc

```diff
@@ -0,0 +1,314 @@
+        d_colorMap->addColorStop(0.6, Qt::green);
+        d_colorMap->addColorStop(0.95, Qt::yellow);
+
+        d_z_colorMap = new QwtLinearColorMap(Qt::darkCyan, Qt::red, QwtColorMap::RGB);
+        d_z_colorMap->addColorStop(0.1, Qt::cyan);
+        d_z_colorMap->addColorStop(0.6, Qt::green);
+        d_z_colorMap->addColorStop(0.95, Qt::yellow);
+
+    } else if (color_map == "indexed") {
+        d_colorMap = new QwtLinearColorMap(Qt::blue, Qt::red, QwtColorMap::Indexed);
+
+        d_z_colorMap = new QwtLinearColorMap(Qt::blue, Qt::red, QwtColorMap::Indexed);
+
+    } else if (color_map == "alpha") {
+        QwtAlphaColorMap* alpha_map = new QwtAlphaColorMap();
+        alpha_map->setColor(Qt::darkBlue);
+        d_colorMap = (QwtLinearColorMap*)alpha_map;
+        d_z_colorMap = (QwtLinearColorMap*)alpha_map;
+    }
+}
+
+void matrix_display::set_interpolation(const std::string& interpolation)
+{
+    //'BilinearInterpolation','NearestNeighbour'
+    if (interpolation == "BilinearInterpolation") {
+        d_data.setResampleMode(QwtMatrixRasterData::ResampleMode::BilinearInterpolation);
+    } else if (interpolation == "NearestNeighbour") {
+        d_data.setResampleMode(QwtMatrixRasterData::ResampleMode::NearestNeighbour);
+    }
+}
+
+void matrix_display::set_x_axis_label(const std::string& x_axis_label)
+{
+    d_x_axis = d_plot->axisWidget(QwtPlot::xBottom);
+    d_x_axis->setTitle(x_axis_label.c_str());
+    d_plot->enableAxis(QwtPlot::xBottom, true);
+}
+
+void matrix_display::set_y_axis_label(const std::string& y_axis_label)
+{
+    d_y_axis = d_plot->axisWidget(QwtPlot::yLeft);
+    d_y_axis->setTitle(y_axis_label.c_str());
+    d_plot->enableAxis(QwtPlot::yLeft, true);
+}
+
+void matrix_display::set_z_axis_label(const std::string& z_axis_label)
+{
+    d_z_axis = d_plot->axisWidget(QwtPlot::yRight);
+    d_z_axis->setTitle(z_axis_label.c_str());
+    d_z_axis->setColorBarEnabled(true);
+    d_z_axis->setColorMap(QwtInterval(d_z_min, d_z_max), d_z_colorMap);
+    d_plot->setAxisScale(QwtPlot::yRight, d_z_min, d_z_max);
+    d_plot->enableAxis(QwtPlot::yRight, true);
+}
+
+
+void matrix_display::save_figure()
+{
+    QPixmap qpix = this->grab();
+
+    QString types = QString(tr("JPEG file (*.jpg);;Portable Network Graphics file "
+                               "(*.png);;Bitmap file (*.bmp);;TIFF file (*.tiff)"));
+
+    QString filename, filetype;
+    QFileDialog* filebox = new QFileDialog(0, "Save Image", "./", types);
+    filebox->setViewMode(QFileDialog::Detail);
+    filebox->setAcceptMode(QFileDialog::AcceptSave);
+    filebox->setFileMode(QFileDialog::AnyFile);
+    if (filebox->exec()) {
+        filename = filebox->selectedFiles()[0];
+        filetype = filebox->selectedNameFilter();
+    } else {
+        return;
+    }
+
+    if (filetype.contains(".jpg")) {
+        qpix.save(filename + ".jpg", "JPEG");
+    } else if (filetype.contains(".png")) {
+        qpix.save(filename + ".png", "PNG");
```

Comment:
Does that mean that if I entered "mypicture.png" as file name, I get "mypicture.png.png" as saved file?

Reply:
I also addressed the same in this issue #6474

Saving it without adding .jpg or other extension made it invisible to the filter and did not showed the earlier saved files,i.e. if a saved one as "abc" (actually a jpeg) and then took another shot the file save dialog did not display any previous saved jpeg files on selecting the filter to jpeg files (which could be inconvenient if I accidentally replaced any previous screenshot with same name )
I checked the saved file in terminal it showed name as abc.jpg not abc.jpg.jpg , I think JPEG is only set as property without extension. Although I checked it in linux only don't know if it shows .jpg.jpg in any other system

Reply:
I'm not sure what you mean with "showed"; what is shown in a graphical file list might or might not include file extensions, I guess?